telemetry/data_slow_query.go

```diff
@@ -0,0 +1,181 @@
+}
+
+// SlowQueryBucket records the statistic information of slow query buckets
+// Buckets:   prometheus.ExponentialBuckets(0.001, 2, 28), // 1ms ~ 1.5days  // defined in metrics/server.go
+type SlowQueryBucket map[string]int
+
+const slowQueryBucketNum = 29 //prometheus.ExponentialBuckets(0.001, 2, 28), and 1 more +Inf
+
+var (
+	// LastSQBInfo records last statistic information of slow query buckets
+	LastSQBInfo SlowQueryBucket
+	// CurrentSQBInfo records current statitic information of slow query buckets
+	CurrentSQBInfo SlowQueryBucket
+)
+
+func getSlowQueryStats(ctx sessionctx.Context) (*slowQueryStats, error) {
+	slowQueryBucket, err := GetSlowQueryBucket(ctx)
+	if err != nil {
+		logutil.BgLogger().Info(err.Error())
+		return nil, err
+	}
+
+	return &slowQueryStats{slowQueryBucket}, nil
+}
+
+// GetSlowQueryBucket genenrates the delta SlowQueryBucket to report
+func GetSlowQueryBucket(ctx sessionctx.Context) (*SlowQueryBucket, error) {
+	// update CurrentSQBInfo first, then gen delta
+	if err := UpdateCurrentSQB(ctx); err != nil {
+		return nil, err
+	}
+	delta := CalculateDeltaSQB()
+	return delta, nil
+}
+
+// UpdateCurrentSQB records current slow query buckets
+func UpdateCurrentSQB(ctx sessionctx.Context) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			switch x := r.(type) {
+			case string:
+				err = errors.New(x)
+			case error:
+				err = x
+			default:
+				err = errors.New("unknown failure")
+			}
+		}
+	}()
+
+	value, err := querySlowQueryMetric(ctx) //TODO: judge error here
+	if err != nil {
+		logutil.BgLogger().Info("querySlowQueryMetric got error")
+		return err
+	}
+
+	if value.Type() != pmodel.ValVector {
+		return errors.New("Prom vector expected, got " + value.Type().String())
+	}
+	promVec := value.(pmodel.Vector)
+	for _, sample := range promVec {
+		metric := sample.Metric
+		bucketName := metric["le"] //hardcode bucket upper bound
+		CurrentSQBInfo[string(bucketName)] = int(sample.Value)
+	}
+	return nil
+}
+
+func querySlowQueryMetric(sctx sessionctx.Context) (result pmodel.Value, err error) {
+	// Add retry to avoid network error.
+	var prometheusAddr string
+	for i := 0; i < 5; i++ {
+		//TODO: the prometheus will be Integrated into the PD, then we need to query the prometheus in PD directly, which need change the quire API
+		prometheusAddr, err = infosync.GetPrometheusAddr()
+		if err == nil || err == infosync.ErrPrometheusAddrIsNotSet {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	if err != nil {
+		return nil, err
+	}
+	promClient, err := api.NewClient(api.Config{
+		Address: prometheusAddr,
+	})
+	if err != nil {
+		return nil, err
+	}
+	promQLAPI := promv1.NewAPI(promClient)
+	promQL := "tidb_server_slow_query_process_duration_seconds_bucket{sql_type=\"general\"}"
+
+	ts := time.Now()
+	// Add retry to avoid network error.
+	ctx := context.TODO() // just use default context
+	for i := 0; i < 5; i++ {
+		result, _, err = promQLAPI.Query(ctx, promQL, ts)
```

Comment:
do we need refresh ts to time.Now during retry~?

Reply:
This ts is used to get the statistic information of slow query bucket from prom. , it won't be so critical to update it when retry. 
Maybe we will miss several slow queries, but it will be count during next report.